oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/R2dbcReactiveOAuth2AuthorizedClientService.java

```diff
@@ -0,0 +1,323 @@
+import org.springframework.security.oauth2.core.OAuth2AccessToken;
+import org.springframework.security.oauth2.core.OAuth2RefreshToken;
+import org.springframework.util.Assert;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.StringUtils;
+
+import io.r2dbc.spi.Row;
+import io.r2dbc.spi.RowMetadata;
+import reactor.core.publisher.Mono;
+
+/**
+ * A R2DBC implementation of {@link ReactiveOAuth2AuthorizedClientService} that
+ * uses a {@link DatabaseClient} for {@link OAuth2AuthorizedClient} persistence.
+ *
+ * <p>
+ * <b>NOTE:</b> This {@code OAuth2AuthorizedClientService} depends on the table
+ * definition described in
+ * "classpath:org/springframework/security/oauth2/client/oauth2-client-schema.sql"
+ * and therefore MUST be defined in the database schema.
+ *
+ * @author Ovidiu Popa
+ * @since 5.4
+ * @see ReactiveOAuth2AuthorizedClientService
+ * @see OAuth2AuthorizedClient
+ * @see DatabaseClient
+ *
+ */
+public class R2dbcReactiveOAuth2AuthorizedClientService implements ReactiveOAuth2AuthorizedClientService {
+	private static final String COLUMN_NAMES =
+			"client_registration_id, " +
+			"principal_name, " +
+			"access_token_type, " +
+			"access_token_value, " +
+			"access_token_issued_at, " +
+			"access_token_expires_at, " +
+			"access_token_scopes, " +
+			"refresh_token_value, " +
+			"refresh_token_issued_at";
+	private static final String TABLE_NAME = "oauth2_authorized_client";
+	private static final String PK_FILTER = "client_registration_id = :clientRegistrationId AND principal_name = :principalName";
+	private static final String LOAD_AUTHORIZED_CLIENT_SQL = "SELECT " + COLUMN_NAMES + " FROM " + TABLE_NAME
+			+ " WHERE " + PK_FILTER;
+	private static final String REMOVE_AUTHORIZED_CLIENT_SQL = "DELETE FROM " + TABLE_NAME + " WHERE " + PK_FILTER;
+	private static final String SAVE_AUTHORIZED_CLIENT_SQL = "INSERT INTO " + TABLE_NAME + " (" + COLUMN_NAMES + ")"+
+			"VALUES (:clientRegistrationId, :principalName, :accessTokenType, :accessTokenValue,"+
+				" :accessTokenIssuedAt, :accessTokenExpiresAt, :accessTokenScopes, :refreshTokenValue,"+
+				" :refreshTokenIssuedAt)";
+	private static final String UPDATE_AUTHORIZED_CLIENT_SQL = "UPDATE " + TABLE_NAME +
+			" SET access_token_type = :accessTokenType, " +
+			" access_token_value = :accessTokenValue, " +
+			" access_token_issued_at = :accessTokenIssuedAt," +
+			" access_token_expires_at = :accessTokenExpiresAt, " +
+			" access_token_scopes = :accessTokenScopes," +
+			" refresh_token_value = :refreshTokenValue, " +
+			" refresh_token_issued_at = :refreshTokenIssuedAt" +
+			" WHERE " +
+			PK_FILTER;
+	private DatabaseClient databaseClient;
+	private ReactiveClientRegistrationRepository clientRegistrationRepository;
+	private BiFunction<OAuth2AuthorizedClient, Authentication, Map<String, Parameter>> authorizedClientParametersMapper;
+	private BiFunction<Row, RowMetadata, OAuth2AuthorizedClientHolder>  authorizedClientRowMapper;
```

Comment:
Hi @jgrandja ,

if we add ```clientRegistrationRepository ``` as a member of the ```OAuth2AuthorizedClientRowMapper``` it means that the return type of this function would be ```Mono<OAuth2AuthorizedClient>```

Reply:
Yes, I see what you're saying. Give it a try and if it works out great otherwise we can keep as-is.

Reply:
Changed the method signature to ``` public Mono<OAuth2AuthorizedClient> apply(Row row, RowMetadata rowMetadata)``` in ```OAuth2AuthorizedClientRowMapper``` class. Tests are passing. 

The only thing that I don't like how it looks is 
```
return (Mono<T>) this.databaseClient.sql(LOAD_AUTHORIZED_CLIENT_SQL)
				.bind("clientRegistrationId", clientRegistrationId).bind("principalName", principalName)
				.map(this.authorizedClientRowMapper::apply)
				.first().flatMap(authorizedClient -> authorizedClient);
```
Because ```first()``` is already returning a Mono, basically the result of he mapper will actually be ```Mono<Mono<OAuth2AuthorizedClient>>``` which means that I have to do a``` flatMap``` to get the underlying ```Mono<OAuth2AuthorizedClient>```